components/i2cdev/i2cdev_master.c

```diff
@@ -0,0 +1,197 @@
+        && a->sda_io_num == b->sda_io_num
+        && a->flags.enable_internal_pullup == b->flags.enable_internal_pullup;
+}
+
+static esp_err_t i2c_setup_device(i2c_dev_t *dev)
+{
+    return i2c_master_bus_add_device(dev->bus, &dev->device_config, &dev->device);
+}
+
+static esp_err_t i2c_setup_port(i2c_dev_t *dev)
+{
+    if (dev->port >= I2C_NUM_MAX) return ESP_ERR_INVALID_ARG;
+
+    esp_err_t res;
+    if (!cfg_equal(&dev->master_bus_config, &states[dev->port].config) || !states[dev->port].installed)
+    {
+        ESP_LOGD(TAG, "Reconfiguring I2C driver on port %d", dev->port);
+        i2c_master_bus_config_t temp;
+        memcpy(&temp, &dev->master_bus_config, sizeof(i2c_master_bus_config_t));
+
+        // Driver reinstallation
+        if (states[dev->port].installed)
+        {
+            if (!dev->bus)
+            {
+                ESP_LOGW(TAG, "The state of I2C port %d is `installed`, but bus handle is NULL)", dev->port);
+            }
+            else
+            {
+                i2c_del_master_bus(dev->bus);
+            }
+            states[dev->port].installed = false;
+        }
+        if ((res = i2c_new_master_bus(&dev->master_bus_config, &dev->bus)) != ESP_OK)
+            return res;
+        states[dev->port].bus = dev->bus;
+        states[dev->port].installed = true;
+
+        memcpy(&states[dev->port].config, &temp, sizeof(i2c_master_bus_config_t));
+        ESP_LOGD(TAG, "I2C driver successfully reconfigured on port %d", dev->port);
+    }
+    return i2c_setup_device(dev);
+}
+
+esp_err_t _i2cdev_init()
+{
+    memset(states, 0, sizeof(states));
+
+#if !CONFIG_I2CDEV_NOLOCK
+    for (int i = 0; i < I2C_NUM_MAX; i++)
+    {
+        states[i].lock = xSemaphoreCreateMutex();
+        if (!states[i].lock)
+        {
+            ESP_LOGE(TAG, "Could not create port mutex %d", i);
+            return ESP_FAIL;
+        }
+    }
+#endif
+
+    return ESP_OK;
+}
+
+esp_err_t _i2cdev_done()
+{
+    for (int i = 0; i < I2C_NUM_MAX; i++)
+    {
+        if (!states[i].lock) continue;
+
+        if (states[i].installed)
+        {
+            SEMAPHORE_TAKE(i);
+            if (!states[i].bus)
+            {
+                ESP_LOGW(TAG, "The state of I2C port %d is `installed`, but bus handle is NULL)", i);
+            }
+            else
+            {
+                i2c_del_master_bus(states[i].bus);
+            }
+            states[i].installed = false;
+            SEMAPHORE_GIVE(i);
+        }
+#if !CONFIG_I2CDEV_NOLOCK
+        vSemaphoreDelete(states[i].lock);
+#endif
+        states[i].lock = NULL;
+    }
+    return ESP_OK;
+}
+
+esp_err_t _i2c_dev_probe(i2c_dev_t *dev, i2c_dev_type_t operation_type)
+{
+    if (!dev) return ESP_ERR_INVALID_ARG;
+
+    SEMAPHORE_TAKE(dev->port);
+
+    esp_err_t res = i2c_setup_port(dev);
+    if (res == ESP_OK)
+    {
+        res = i2c_master_probe(dev->bus, dev->device_config.device_address, CONFIG_I2CDEV_TIMEOUT);
+    }
+
+    SEMAPHORE_GIVE(dev->port);
```

Comment:
This is what I meant with the repetition, wouldn't the code be much clearly arranged to only interchange the calls to the old and new i2c drivers (using preprocessor statements)?

Reply:
I am inclined to avoid macro because of readability when the change is not trivial. The reason of separating old and new is that, you can modify both as long as the interface (and other expected behaviors) are kept. Others might not agree, though.

Reply:
That does make sense, but it may introduce two versions of the "same" logic. My opinion is that it may be less readable, but safer to make changes to (because you won't have to make a change twice, removing the possibility of making a mistake)